examples/09_Kotlin_JS/06_HtmlBuilder.md

```diff
@@ -47,25 +43,61 @@ fun main(args: Array<String>) {
         }
     }
     //sampleEnd
+    println(result)
+}
+
+interface Element {
+    fun render(builder: StringBuilder, indent: String)
+}
 
-    document.body!!.appendChild(result.element)
+class TextElement(val text: String) : Element {
+    override fun render(builder: StringBuilder, indent: String) {
+        builder.append("$indent$text\n")
+    }
 }
 
-abstract class Tag(val name: String) {
-    val element = document.createElement(name)
-    protected fun <T : Tag> initTag(tag: T, init: T.() -> Unit): T {
+@DslMarker
+annotation class HtmlTagMarker
+
+@HtmlTagMarker
+abstract class Tag(val name: String) : Element {
+    val children = arrayListOf<Element>()
+    val attributes = hashMapOf<String, String>()
+
+    protected fun <T : Element> initTag(tag: T, init: T.() -> Unit): T {
         tag.init()
-        element.appendChild(tag.element)
+        children.add(tag)
         return tag
     }
+
+    override fun render(builder: StringBuilder, indent: String) {
+        builder.append("$indent<$name${renderAttributes()}>\n")
+        for (c in children) {
+            c.render(builder, indent + "  ")
+        }
+        builder.append("$indent</$name>\n")
+    }
+
+    private fun renderAttributes(): String {
+        val builder = StringBuilder()
+        for ((attr, value) in attributes) {
+            builder.append(" $attr=\"$value\"")
+        }
+        return builder.toString()
+    }
+
+    override fun toString(): String {
+        val builder = StringBuilder()
+        render(builder, "")
+        return builder.toString()
+    }
 }
 
 abstract class TagWithText(name: String) : Tag(name) {
     operator fun String.unaryPlus() {
-        element.appendChild(document.createTextNode(this + " "))
+        children.add(TextElement(this))
     }
 }
-
 class HTML() : TagWithText("html") {
```

Comment:
The version with your update does not parse \<ul> and \<li> tags. 
Please add this function into HTML:
`fun ul(init: UL.() -> Unit) = initTag(UL(), init)`
and see comments below.

```suggestion
class HTML() : TagWithText("html") {
    fun ul(init: UL.() -> Unit) = initTag(UL(), init)
```

Reply:
This does not look right as HTML consists of **head** and **body** while the **ul** tag goes inside the **body**. I already included the **ul** fun into the abstract class BodyTag.
Unfortunately, I have accidentally merged the change, and now do not see how to revert this commit.

Reply:
Yes, you're completely right. My bad, I haven't checked the whole file in your branch, just the changes. Sorry.
Let's just remove the redundant declarations in a separate commit. I'll squash all this PR into one commit with the right corrections when merging.
Please remove:
- `fun ul()` from `HTML` class
- second declarations of `ul` and `li` classes.

Reply:
Please note that before submitting the PR I ensured that the code compiles and produces correct results.
I reverted all new commits and re-applied the first one with a meaningful name.